src/index.js

```diff
@@ -768,58 +769,94 @@ class Serverless_WSO2_APIM {
+          await deploySwaggerSpecs(wso2APIM, api.apiName, apiId, apiDef)
+        }
+      } catch (err) {
+        this.serverless.cli.log(`Creating / Updating ${apiDef.name}.. NOT OK, proceeding further.`);
+
+        // ? do not retry in case of errors in the deployment
+        return;
+      }
+
+      // ? Only check if the apiDef was properly updated when it is an update operation
+      if (isCreatingApiDef) {
+        return;
+      }
+
+      const apiDefIsUpdated = await wso2apim.checkApiDefIsUpdated(
+        wso2APIM,
+        this.cache.accessToken,
+        api.apiId,
+        apiDef,
+      );
+
+      if (!apiDefIsUpdated) {
+        throw new Error('Api definition is outdated');
+      }
+    };
+
     try {
       // By calling listAPIDefs(), we are re-collecting the current deployment status in WSO2 API Manager into this.cache.deploymentStatus
       await this.listAPIDefs();
       // Loop thru this.cache.deploymentStatus array
       // Create API definitions, if they do not exist
       // Update API definitions, if they exist
       for (const [i, api] of this.cache.deploymentStatus.entries()) {
-        try {
-          let apiId = api.apiId;
-          //Update
-          if (api.apiId !== undefined) {
-            // this.serverless.cli.log(pluginNameSuffix + "Updating " + api.apiName + " (" + api.apiId + ")..");
-            await wso2apim.updateAPIDef(
-              wso2APIM,
-              this.cache.accessToken,
-              apiDefs[i],
-              api.apiId
-            );
-            this.serverless.cli.log(
-              pluginNameSuffix +
-                'Updating ' +
-                api.apiName +
-                ' (' +
-                api.apiId +
-                ').. OK'
-            );
-          }
-          //Create
-          else {
-            // this.serverless.cli.log(pluginNameSuffix + "Creating " + api.apiName + "".."");
-            const result = await wso2apim.createAPIDef(
-              wso2APIM,
-              this.cache.accessToken,
-              apiDefs[i]
-            );
-            apiId = result.apiId;
-            this.serverless.cli.log(
-              pluginNameSuffix + 'Creating ' + api.apiName + '.. OK'
-            );
-          }
-
-          // now update the swagger spec of the API
-          this.serverless.cli.log(`${pluginNameSuffix}Upserting Swagger spec for ${api.apiName}..`);
-          await wso2apim.upsertSwaggerSpec(wso2APIM, this.cache.accessToken, apiId, apiDefs[i].swaggerSpec);
-          this.serverless.cli.log(`${pluginNameSuffix}Upserting.. OK`);
-        } catch (err) {
-          this.serverless.cli.log(
-            'Creating / Updating ' +
-              `${apiDefs[i].name}` +
-              '.. NOT OK, proceeding further.'
-          );
-        }
+        // ? seconds sleeping on each retry: 2, 4, 8, 16, 32, 64
+        await retryWithExponentialBackoff(({ attempt }) => deployApiDefAndUpsertSwagger(wso2APIM, api, apiDefs[i], attempt), {
```

Comment:
question: should we add a log here on every fail/retry? to help us understand possible errors in the future?

Reply:
We can realize that it was retried because the update message is repeated...
<img width="750" alt="Screenshot 2024-02-02 at 14 05 53" src="https://github.com/ramgrandhi/serverless-wso2-apim/assets/34343596/0231a761-7858-4c07-a83a-24a622f2ebb8">